docs/proposals/20210335-load-balancer-provider.md

```diff
@@ -0,0 +1,999 @@
+   targets: (List of members matching label selector)
+   activeTargets: (List of active members)
+```
+#### TODO
+
+- define conditions
+
+#### Watches:
+
+- own resource
+- Kind referenced by LB selector (to identify changes to "disabled" annotation)
+- (???) LoadBalancer resources associated with own resource
+
+#### Reconciliation
+TODO: Attempt to load InfraCluster and switch to alternate behaviour when != AWSCluster.
+- If Spec.Endpoint is not defined:
+    - Create new LB using provider defaults for endpoint, populate Spec.Endpoint
+- Lookup related LoadBalancer resource:
+    - For any matching resources:
+        - resource is not deleting:
+            - `loadbalancer.cluster.x-k8s.io/disabled` not present:
+                - resource not present in load balancer:
+                    - add resource to load balancer config
+                    - set `loadbalancer.cluster.x-k8s.io/status` to 'adding'
+                    - requeue resource appropriately
+                - resource present in load balancer, but not active:
+                    - update `loadbalancer.cluster.x-k8s.io/status` appropriately for resource
+                    - requeue resource appropriately
+                - resource present and active in load balancer:
+                    - set `loadbalancer.cluster.x-k8s.io/status` to 'active'
+            - `loadbalancer.cluster.x-k8s.io/disabled` is present:
+                - resource active in load balancer:
+                    - start removal/disabling of resource from lb
+                    - set `loadbalancer.cluster.x-k8s.io/status` to 'disabling'
+                    - requeue resource appropriately
+                - resource is being removed/disabled from lb:
+                    - update `loadbalancer.cluster.x-k8s.io/status` appropriately for resource
+                    - requeue resource appropriately
+                - resource is removed/disabled from the load balancer:
+                    - set `loadbalancer.cluster.x-k8s.io/status` to 'disabled'
+        - resource is deleting:
+            - resource active in load balancer:
+                - start removal of resource from lb:
+                - set `loadbalancer.cluster.x-k8s.io/status` to 'removing'
+                - requeue resource appropriately
+            - resource is being removed from lb:
+                - update `loadbalancer.cluster.x-k8s.io/status` appropriately for resource
+                - requeue resource appropriately
+            - resource is removed from load balancer:
+                - set `loadbalancer.cluster.x-k8s.io/status` to 'removed'
+
+- Ensure there are no resources configured in the load balancer that do not match the selector
+	
+
+LB provider registers pre-drain or pre-term hook on machines that are LoadBalancerMembers
+CP provider issues machine delete
+LB provider processes hook, quiesces, removes its hook annotation
+Machine is deleted
+
```

Comment:
> I know WIP :). I think we need an example for non-control-plane loadbalancers to think those cases through.
> Some questions:
> 
> * How would non-control-plane loadbalancers be "connected" to clusters? I assume similar to MachineDeployments today?

My thought is that they would be somewhat disjoint, connected to the Cluster through the `Spec.ClusterName` field, at least to start. From the perspective of the Cluster, we could handle it similar to how we handle other "descendants" of the Cluster, such as MachineDeployments, MachineSets, and Machines that have 1:many relationships.

> * Is it possible to reference `targetKinds` of multiple clusters in non-control-plane loadbalancers?

Is there a use case where that would be desirable? I'd probably lean towards 'no' without a really compelling use case where it would be needed.

Reply:
>> Is it possible to reference targetKinds of multiple clusters in non-control-plane loadbalancers?
> Is there a use case where that would be desirable? I'd probably lean towards 'no' without a really compelling use case where it would be needed.

I agree. I was just wondering. As far as I know the OpenStack ccm allows loadbalancer re-use between different Kubernetes clusters. If nobody really needs it, I wouldn't support it as it adds additional complexity.